AOC [23.07](https://adventofcode.com/2023/day/7) Camel Cards

Poker variation: 
* cards from A, K, Q, J, T, 9 - 2.
* ranks:
 * 5 of a kind
 * 4 of a kind
 * full house: 3 of a kind + pair
 * three of a kind
 * two pair
 * one pair
 * high card

In case of tied hands: look at cards in hand. First card higher = winning hand.
* 33332 > 2AAAA b/c 3>2
* 77888 < 88877 

Given input of hands and bids, put them in rank order (1 = low, N = high). Multply bids by rank. Tally product.

Solution:
1. create a function to compare two hands and rank one above the other
1. sort the list using list.sort and a lambda function that compares the two hands


In [58]:
testData = '''32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483'''

data = [tuple(x.split()) for x in testData.split('\n')]
data


[('32T3K', '765'),
 ('T55J5', '684'),
 ('KK677', '28'),
 ('KTJJT', '220'),
 ('QQQJA', '483')]

In [ ]:
# https://codereview.stackexchange.com/questions/144551/find-and-display-best-poker-hand

import itertools

def numeric_ranks(cards):
    suits = get_suits(cards)
    face_numbers = {'A': 14, 'J': 11, 'Q': 12, 'K': 13}
    for index, card in enumerate(cards):
        rank = card[0:-1]
        try: int(rank)
        except:  # Rank is a letter, not a number
            cards[index] = str(face_numbers[rank])+suits[index]
    return cards

def get_ranks(cards):
    cards = numeric_ranks(cards) # Convert rank letters to numbers (e.g. J to 11)
    return [int(card[0:-1]) for card in cards]

def get_suits(cards):
    return [card[-1] for card in cards]

def evaluate_hand(hand):
    hand = numeric_ranks(hand)
    ranks = get_ranks(hand)
    suits = get_suits(hand)
    if len(set(hand)) < len(hand) or max(ranks) > 14 or min(ranks) < 1:  # There is a duplicate
        return 'Invalid hand'
    if isconsecutive(ranks):  # The hand is a type of straight
        if all_equal(suits):  # Hand is a flush
            if max(ranks) == 14:  # Highest card is an ace
                return 'Royal flush'
            return 'Straight flush'
        return 'Straight'
    if all_equal(suits):
        return 'Flush'
    total = sum([ranks.count(x) for x in ranks])
    hand_names = {
        17: 'Four of a kind',
        13: 'Full house',
        11: 'Three of a kind',
        9: 'Two pair',
        7: 'One pair',
        5: 'High card'
    }
    return hand_names[total]

def all_equal(lst):
    return len(set(lst)) == 1


In [101]:
import itertools

def numeric_ranks(cards): # input string of five characters; return list of 5 integers representing the cards
    face_numbers = {'A': 14, 'J': 11, 'Q': 12, 'K': 13, 'T':10}
    intCards = []
    for card in cards: # look at each card and assign numberic value
        if card.isdigit():
            intCards.append(int(card))
        else:
            intCards.append(face_numbers[card])
    return intCards

def evaluate_hand(ranks):
    #ranks = numeric_ranks(hand)
    total = [ranks.count(x) for x in set(ranks)]
    #print(total)
    if 5 in total: return 7 # five of a kind
    elif 4 in total: return 6
    elif (3 in total) and (2 in total): return 5
    elif 3 in total: return 4
    elif len([i for i in total if i == 2]) == 2: return 3
    elif 2 in total: return 2
    else: return 1

valued_hands = []
hands = [(numeric_ranks(x[0]),int(x[1])) for x in data]

for hand in hands:
    valued_hands.append(hand + (evaluate_hand(hand[0]),))


[([3, 2, 10, 3, 13], 765, 2), ([10, 5, 5, 11, 5], 684, 4), ([13, 13, 6, 7, 7], 28, 3), ([13, 10, 11, 11, 10], 220, 3), ([12, 12, 12, 11, 14], 483, 4)]


In [111]:
hands

[([3, 2, 10, 3, 13], 765),
 ([10, 5, 5, 11, 5], 684),
 ([13, 13, 6, 7, 7], 28),
 ([13, 10, 11, 11, 10], 220),
 ([12, 12, 12, 11, 14], 483)]

In [126]:
def bubbleSort(arr):
    n = len(arr)
    swapped = False
    for i in range(n-1):
        for j in range(0, n-i-1):
            if arr[j][2] > arr[j + 1][2]: # look at ranked scored
                swapped = True
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
            elif arr[j][2] == arr[j+1][2]: # if scores are equal, look at card values
                for k in range(4):
                    if arr[j][0][k] > arr[j+1][0][k]:
                        swapped = True
                        arr[j], arr[j+1] =  arr[j+1], arr[j] 
                        break
                    else: continue
        if not swapped:
            break
    return arr

sorted_arr = bubbleSort(valued_hands)

#print(sorted_arr)

tally = 0
for i in range(0,len(sorted_arr)):
    tally += (i+1) * sorted_arr[i][1]

print(tally)

[([3, 2, 10, 3, 13], 765, 2), ([13, 13, 6, 7, 7], 28, 3), ([13, 10, 11, 11, 10], 220, 3), ([10, 5, 5, 11, 5], 684, 4), ([12, 12, 12, 11, 14], 483, 4)]
6632


OK,got the test data to work correctly. Let's try it on the real data.

In [206]:
def numeric_ranks(cards): # input string of five characters; return list of 5 integers representing the cards
    face_numbers = {'A': 14, 'J': 11, 'Q': 12, 'K': 13, 'T':10}
    intCards = []
    for card in cards: # look at each card and assign numberic value
        if card.isdigit():
            intCards.append(int(card))
        else:
            intCards.append(face_numbers[card])
    return intCards

def evaluate_hand(ranks):
    #ranks = numeric_ranks(hand)
    total = [ranks.count(x) for x in set(ranks)]
    #print(total)
    if 5 in total: return 7 # five of a kind
    elif 4 in total: return 6
    elif (3 in total) and (2 in total): return 5
    elif 3 in total: return 4
    elif len([i for i in total if i == 2]) == 2: return 3
    elif 2 in total: return 2
    else: return 1

def compareCards(card1,card2): # compare two cards; if card 1 indexes > card 2, return True
    swap = False
    if card1[0] > card2[0] : return True
    for i in range(1,len(card)):
        if (card1[i-1] == card2[i-1]) and (card1[i] > card2[i]):
            swap = True
            break
    if swap: return True
    else: return False

def bubbleSort(arr):
    n = len(arr)
    swapped = False
    for i in range(n-1):
        for j in range(0, n-i-1):
            if arr[j][2] > arr[j + 1][2]: # look at ranked scored
                swapped = True
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
            elif arr[j][2] == arr[j+1][2]: # if scores are equal, look at card values
                if compareCards(arr[j][0],arr[j+1][0]):
                    swapped = True
                    arr[j], arr[j + 1] = arr[j + 1], arr[j]
        if not swapped:
            break
    return arr

with open ('2307input.txt') as f_in:
    data = [tuple(x.split()) for x in f_in.read().split('\n')]
#data = [tuple(x.split()) for x in testData.split('\n')]

hands = [(numeric_ranks(x[0]),int(x[1])) for x in data] # convert the hands into tuples of numbered hands + wager ([3,2,10,3,13], 765)

valued_hands = []
for hand in hands: # evaluate the hands and create new list of tuples with the value as third element ([x,x,x,x,x],wager,value)
    valued_hands.append(hand + (evaluate_hand(hand[0]),))

sorted_hands = bubbleSort(valued_hands)


tally = 0
for i in range(0,len(sorted_hands)):
    #print(f"{i+1}, {sorted_hands[i][1]}")
    tally += (i+1) * sorted_hands[i][1]

print(tally)


250988697


This value is too low. I think my sort routine isn't working properly. Let's dump it and use the sort routines in the python library. Rather than doing one sort, let's split the sort process into two steps:
1. sort the list of hands by their numeric values --> this will create 7 groups of cards (values 1, 2, 3)
1. sort each group of card by their lists -- the python sort is smart enough to sort them pairwise

And now that I've written this, I realize I can skip the first sort. If I just iterate through the list, select those with hand value = 1, then 2 etc., I just need to call the second sort!

In [223]:
def numeric_ranks(cards): # input string of five characters; return list of 5 integers representing the cards
    face_numbers = {'A': 14, 'J': 11, 'Q': 12, 'K': 13, 'T':10}
    intCards = []
    for card in cards: # look at each card and assign numberic value
        if card.isdigit():
            intCards.append(int(card))
        else:
            intCards.append(face_numbers[card])
    return intCards

def evaluate_hand(ranks):
    #ranks = numeric_ranks(hand)
    total = [ranks.count(x) for x in set(ranks)]
    #print(total)
    if 5 in total: return 7 # five of a kind
    elif 4 in total: return 6
    elif (3 in total) and (2 in total): return 5
    elif 3 in total: return 4
    elif len([i for i in total if i == 2]) == 2: return 3
    elif 2 in total: return 2
    else: return 1

def compareCards(card1,card2): # compare two cards; if card 1 indexes > card 2, return True
    swap = False
    if card1[0] > card2[0] : return True
    for i in range(1,len(card)):
        if (card1[i-1] == card2[i-1]) and (card1[i] > card2[i]):
            swap = True
            break
    if swap: return True
    else: return False

with open ('2307input.txt') as f_in:
    data = [tuple(x.split()) for x in f_in.read().split('\n')]
#data = [tuple(x.split()) for x in testData.split('\n')]

hands = [(numeric_ranks(x[0]),int(x[1])) for x in data] # convert the hands into tuples of numbered hands + wager ([3,2,10,3,13], 765)

valued_hands = []
for hand in hands: # evaluate the hands and create new list of tuples with the value as third element ([x,x,x,x,x],wager,value)
    valued_hands.append(hand + (evaluate_hand(hand[0]),))

#sorted_hands = sorted(valued_hands, key = lambda index: index[2]) # sort by the hand value; this will create 7 "piles" of hands

finalList = []
for i in range(1,8):
    tempList = []
    for k in range(len(valued_hands)):
        if valued_hands[k][2] == i: tempList.append(valued_hands[k])
    finalList.extend(sorted(tempList,key = lambda index: index[0]))

tally = 0
for i in range(0,len(finalList)):
    #print(f"{i+1}, {finalList[i]}")
    tally += (i+1) * finalList[i][1]

print(tally)


250946742


250946742 is correct! On to part 2!

In part 1, J = Jack = 11.

In part 2, J = Joker = wild card -- it can be any card that would maximize the hand's value.

In determining ties between cards, it is the lowest value card (lower than $2$).

Solution:
1. I can change the procedure to turn the letters into number and make J = 1
1. Keep the same valuation, but 
    1. remove the 1s for the initial count & valuation
    1. Count the number of 1s. For each 1, increase the rank by appropriately.


In [277]:
def numeric_ranks(cards): # input string of five characters; return list of 5 integers representing the cards
    face_numbers = {'A': 14, 'J': 1, 'Q': 12, 'K': 13, 'T':10}
    intCards = []
    for card in cards: # look at each card and assign numberic value
        if card.isdigit():
            intCards.append(int(card))
        else:
            intCards.append(face_numbers[card])
    return intCards

def evaluate_hand(card_list):
    ones = card_list.count(1) # how many 1s are in hand? (0 to 5)
    high_card_dict = {0:1,1:2,2:4,3:6,4:7,5:7}
    three_kind_dict = {0:4,1:6,2:7}
    pair_dict = {0:2,1:4,2:6,3:7}

    card_set = set(card_list) # create my set of card values, but get rid of 1s (Jokers)
    card_set.discard(1)

    total = [card_list.count(x) for x in card_set]
    #print(total)
    if 5 in total: return 7 # five of a kind
    elif (3 in total) and (2 in total): return 5 # 5 cards accounted for, no Js
    elif 4 in total: return 6 + ones # four of a kind = 6, five of a kind = 7
    elif 3 in total: return three_kind_dict[ones] # 3 of a kind = 4; 4 of a kind = 6; 5 of a kind = 7
    elif len([i for i in total if i == 2]) == 2: return 3 + ones*2 # two pair = 2; full house = 5
    elif 2 in total: return pair_dict[ones] # pair = 2; 3kind = +2; 3kind +4; 5 kind+5
    else: return high_card_dict[ones] # high = 1; pair = 2; 3kind = 4; 4kind = 6; 5kind = 7  

with open ('2307input.txt') as f_in:
    data = [tuple(x.split()) for x in f_in.read().split('\n')]
#data = [tuple(x.split()) for x in testData.split('\n')]

hands = [(numeric_ranks(x[0]),int(x[1])) for x in data] # convert the hands into tuples of numbered hands + wager ([3,2,10,3,13], 765)

valued_hands = []
for hand in hands: # evaluate the hands and create new list of tuples with the value as third element ([x,x,x,x,x],wager,value)
    valued_hands.append(hand + (evaluate_hand(hand[0]),))

finalList = []
for i in range(1,8):
    tempList = []
    for k in range(len(valued_hands)):
        if valued_hands[k][2] == i: tempList.append(valued_hands[k])
    finalList.extend(sorted(tempList,key = lambda index: index[0]))

tally = 0
for i in range(0,len(finalList)):
    #print(f"{i+1}, {finalList[i]}")
    tally += (i+1) * finalList[i][1]

print(f"{tally:,}, {tally}")


251,824,095, 251824095


Test data works (5909). Try with production data.

First run  245134884 is too low.
Second run 245100059 is lower than the first one!

Third run: added dictionaries for different hand values.
251824095 is the correct answer!